docs/DeferredSignals.md

```diff
@@ -0,0 +1,246 @@
+   - We currently just leak stack for every guest signal handler that long jumps out of the signal frame.
+      - Long term this would exhaust our stack and then crash.
+      - Test with a second guest thread where our host will only have an 8MB stack instead of the 128MB primary stack.
+      - See issue #2487
+## Expectations and considerations
+### What happens with a race condition with the refcounter?
+There are two edges to this problem. The incrementing edge and the decrementing edge that must be considered.
+
+#### Incrementing edge
+This is the most problematic edge. This takes three instructions (one on x86) to increment the ref counter. If a signal is received between the load
+and store then this theoretically could result in a tear on the refcounter. In actual practice this is a real tear but doesn't cause any problems.
+
+The reasoning for this is that FEX isn't in the "uninterruptible" section until that reference counter has been stored, so FEX will handle the signals
+immediately at that point, return to this code location, and then increment the counter. In particular, once returning to the code location the
+refcounter will be the original value loaded. So even though it is a tear, it's one that doesn't cause issues since it is all thread local.
+
+#### Decrementing edge
+This edge is far less problematic to understand compared to the incrementing edge. Signals will get deferred entirely until the store instruction (If
+storing zero), so FEX will always return to the code region and finish the decrement.
+
+If FEX receives a signal after the decrement store has completed but /before/ the page faulting store has occured, then FEX will start processing the
+signal immediately. At which point the fault page will have either RW or NONE permission. FEX will then likely hit another "uninterruptible" code
+section which will complete the store to the fault page.
+ - RW permission if it hadn't received another signal in the uninterruptible section
+ - NONE permission if it did receive a signal previously
+
+RW permission has no problems, it will continue as normal.
+
+NONE will get captured by the fault handler, the fault handler will determine that there was no deferred signals, and set the fault page back to RW
+permissions and continue execution safely.
+
+## Execution examples
+### No signal
+This is a simple example because nothing happens.
+```
+- <Enter Deferred region - 3 instructions>
+! Compiling JIT Code
++ <Exit deferred region - 5 instructions>
+```
+
+### Signal outside of region
+This is simple because the JIT just handles it.
+```
+! In JIT code
+# Signal received
+# Guest Signal handler called
+# JIT jumps to guest signal handler
+# Hopefully guest calls rt_sigreturn instead of long jumping out.
+```
+
+### Synchronous signal in JIT
+Deferred signals don't affect anything here because only asynchronous signals get affected.
+* State reconstruction problems aren't discussed here.
```

Comment:
I don't think mentioning what's not discussed adds much here.

Reply:
Removed. Resolved.